### PR TITLE
Added an option to ReadsSparkSink specifying whether to sort the reads on output. 

### LINKED ISSUE #4859: Make ReadsSparkSource Sorting configurable
Currently ReadsSparkSource automatically sorts reads before writing them out according to the sort order of the header (or without sorting if the header sort order is unrecognized). There should be a way to avoid this for tools where the reads RDD ordering can be guarinteed at write time, or in which we don't want to change the sort ordering anyway like PrintReadsSpark.

### PR DESCRIPTION
**Pull request by jamesemery**
Fixes #4859

Depends on #4545

### CODE REVIEW

**Comment by jamesemery** [comment]
@lbergelson Can we revisit this branch? We need to clean up the current sorting behavior as it is totally inconsistent and confusing. Here was an attempt to clean up the behavior so it will be more uniform.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java [hunk 1]
```diff
@@ -166,7 +166,7 @@ public SparkHeaderlessCRAMOutputFormat() {
     public static void writeReads(
             final JavaSparkContext ctx, final String outputFile, final String referenceFile, final JavaRDD<GATKRead> reads,
             final SAMFileHeader header, ReadsWriteFormat format) throws IOException {
-        writeReads(ctx, outputFile, referenceFile, reads, header, format, 0, null);
+        writeReads(ctx, outputFile, referenceFile, reads, header, format, 0, null,format==ReadsWriteFormat.SINGLE);
```
Comment (tomwhite) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: Nit: missing space before comma

File: src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/PrintReadsSparkIntegrationTest.java [hunk 2]
```diff
@@ -242,6 +242,26 @@ public void testUnSorted() throws Exception {
         SamAssertionUtils.assertSamsEqual(outBam, inBam);
     }
 
+    @Test( groups = "spark")
+    public void testUnSorted() throws Exception {
```
Comment (tomwhite) [inline src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/PrintReadsSparkIntegrationTest.java]: This test looks like it's not finished? It shares a name with an existing test, and `print_reads.mismatchedHeader.sam` is unused.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java (hunk 1 above)
Comment (lbergelson) [reply src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: Lets add a comment about how this is here to maintain old weird behavior.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java [hunk 3]
```diff
@@ -201,19 +202,20 @@ public static void writeReads(
         // SAMRecords, this will effectively be a no-op. The SAMRecords will be headerless
         // for efficient serialization.
         final JavaRDD<SAMRecord> samReads = reads.map(read -> read.convertToSAMRecord(null));
+        final JavaRDD<SAMRecord> readsToUse = sortReadsToHeader ? sortSamRecordsToMatchHeader(samReads, header, numReducers) : samReads;
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: maybe readsToUse -> readsToOutput would be clearer?

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBwaSpark.java [hunk 4]
```diff
@@ -180,7 +180,7 @@ private void writeBam(final JavaRDD<GATKRead> reads, final String inputBamPath,
         try {
             ReadsSparkSink.writeReads(ctx, outputPath, bwaArgs.referencePath, reads, header,
                     shardedOutput ? ReadsWriteFormat.SHARDED : ReadsWriteFormat.SINGLE,
-                    PSUtils.pathseqGetRecommendedNumReducers(inputBamPath, numReducers, getTargetPartitionSize()), shardedPartsDir);
+                    PSUtils.pathseqGetRecommendedNumReducers(inputBamPath, numReducers, getTargetPartitionSize()), shardedPartsDir, true);
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBwaSpark.java]: why is this getting an extra true here?  I don't see any changes to `PSUtils.pathseqGetRecommendedNumReducers`

**Comment by jamesemery** [comment]
@lbergelson @droazen This branch has been brought to speed with master now, I need somebody to review it however....

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java [hunk 5]
```diff
@@ -52,7 +52,9 @@
     public static void writeReads(
             final JavaSparkContext ctx, final String outputFile, final String referenceFile, final JavaRDD<GATKRead> reads,
             final SAMFileHeader header, ReadsWriteFormat format) throws IOException {
-        writeReads(ctx, outputFile, referenceFile, reads, header, format, 0, null);
+        // NOTE, we must include 'format==ReadsWriteFormat.SINGLE' to preserve the old default behavior for writing spark output
+        // which would not sort the bam if outputting to ReadsWriteFormat.SINGLE. Please use the overload for different sroting behavior.
+        writeReads(ctx, outputFile, referenceFile, reads, header, format, 0, null, format==ReadsWriteFormat.SINGLE);
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: I'd like to remove this particular wierdness.  How hard is it to just eliminate this behavior?

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java [hunk 6]
```diff
@@ -66,11 +68,12 @@ public static void writeReads(
...
      *                    should be used.
      * @param outputPartsDir directory for temporary files for SINGLE output format, should be null for default value of filename + .output
+     * @param sortReadsToHeader if true, the writer will perform a sort of reads according to the sort order of the header before writing
      */
     public static void writeReads(
             final JavaSparkContext ctx, final String outputFile, final String referenceFile, final JavaRDD<GATKRead> reads,
-            final SAMFileHeader header, ReadsWriteFormat format, final int numReducers, final String outputPartsDir) throws IOException {
-        writeReads(ctx, outputFile, referenceFile, reads, header, format, numReducers, outputPartsDir, true, true);
+            final SAMFileHeader header, ReadsWriteFormat format, final int numReducers, final String outputPartsDir, final boolean sortReadsToHeader) throws IOException {
+        writeReads(ctx, outputFile, referenceFile, reads, header, format, numReducers, outputPartsDir, true, true, true);
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: this seems like a bug... you added a boolean parameter to the method and then set it to be true always.

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java [hunk 7]
```diff
@@ -48,9 +48,6 @@ private void shutdownMiniCluster() {
     public Object[][] loadReadsBAM() {
         return new Object[][]{
                 {testDataDir + "tools/BQSR/HiSeq.1mb.1RG.2k_lines.bam", "ReadsSparkSinkUnitTest1", null, ".bam", true, true},
-                {testDataDir + "tools/BQSR/HiSeq.1mb.1RG.2k_lines.bam", "ReadsSparkSinkUnitTest1", null, ".bam", true, false}, // write BAI, don't write SBI
```
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java]: Did you mean to delete these test cases?

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java [hunk 8]
```diff
@@ -149,7 +146,7 @@ private void assertSingleShardedWritingWorks(String inputBam, String referenceFi
         JavaRDD<GATKRead> rddParallelReads = readSource.getParallelReads(inputBam, referenceFile);
         SAMFileHeader header = readSource.getHeader(inputBam, referenceFile);
 
-        ReadsSparkSink.writeReads(ctx, outputPath, referenceFile, rddParallelReads, header, ReadsWriteFormat.SINGLE, 0, outputPartsPath, writeBai, writeSbi);
+        ReadsSparkSink.writeReads(ctx, outputPath, referenceFile, rddParallelReads, header, ReadsWriteFormat.SINGLE, 0, outputPartsPath, writeBai, writeSbi, true);
```
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java]: you should have a test case for writing unsorted reads in this class I think.

File: src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/PrintReadsSparkIntegrationTest.java [hunk 9]
```diff
@@ -238,6 +238,26 @@ public void testUnSorted() throws Exception {
...
+    public void testUnSortedRoundTrip() throws Exception {
+        // This is a technically incorrectly sam with a header indicating that it is coordinate sorted when it is actually
+        // queryname sorted. If the ordering is the same after PrintReadsSpark then it means we aren't automatically sorting the output.
+        final File inBam = new File(getTestDataDir(), "print_reads.mismatchedHeader.sam");
+        try (ReadsDataSource ds = new ReadsDataSource(inBam.toPath())){
+            Assert.assertEquals(ds.getHeader().getSortOrder(), SAMFileHeader.SortOrder.coordinate);
+        }
+        final File outBam = GATKBaseTest.createTempFile("print_reads", ".bam");
+        ArgumentsBuilder args = new ArgumentsBuilder();
+        args.add("--" + StandardArgumentDefinitions.INPUT_LONG_NAME);
```
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/PrintReadsSparkIntegrationTest.java]: We have convenience methods for this..

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java (hunk 5 above)
Comment (lbergelson) [reply src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: I think the only remaining use is in BwaSpark and that can be changed to just always sort I think.

Comment (lbergelson) [reply src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: Or to take a parameter to sort I guess.

Comment (lbergelson) [reply src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: Either way...

**Comment by jamesemery** [comment]
@lbergelson Responded to your comments, I left the behavior in bwaSpark for sharded writing as I suspect that was the specific use case for which it was introduced. Back to you...

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java [hunk 10]
```diff
@@ -52,7 +52,9 @@
     public static void writeReads(
             final JavaSparkContext ctx, final String outputFile, final String referenceFile, final JavaRDD<GATKRead> reads,
             final SAMFileHeader header, ReadsWriteFormat format) throws IOException {
-        writeReads(ctx, outputFile, referenceFile, reads, header, format, 0, null);
+        // NOTE, we must include 'format==ReadsWriteFormat.SINGLE' to preserve the old default behavior for writing spark output
+        // which would not sort the bam if outputting to ReadsWriteFormat.SINGLE. Please use the overload for different sorting behavior.
+        writeReads(ctx, outputFile, referenceFile, reads, header, format, 0, null, true);
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: If you think this behavior is important to bwaspark, push it down there.  This is going to bite us in the future if we leave it in.

Comment (lbergelson) [reply src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: Oh, you did, you just forgot the comment.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/bwa/BwaSpark.java [hunk 11]
```diff
@@ -74,7 +74,7 @@ protected void runTool(final JavaSparkContext ctx) {
             }
             try {
                 ReadsSparkSink.writeReads(ctx, output, null, reads, bwaEngine.getHeader(),
-                                            shardedOutput ? ReadsWriteFormat.SHARDED : ReadsWriteFormat.SINGLE);
+                                            shardedOutput ? ReadsWriteFormat.SHARDED : ReadsWriteFormat.SINGLE, getRecommendedNumReducers(), shardedPartsDir, shardedOutput);
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/bwa/BwaSpark.java]: Maybe comment here so people know that something weird is going on?

**Comment by jamesemery** [comment]
@lbergelson Added a comment to BwaSpark. Can this be merged?

**Pull request merged**